release-train-model: write down the first version (#60)
pingcap/tidb-dev-guide: src/project-management/release-train-model.md

# Release Train Model

> 

Author: "We" are TiDB developers and I don't think it is misleading. Postpone until others complain also.

## What is release train model?

Before v5.0, TiDB releases a major version every year or every half year.

However, the longer a development cycle is, the more features it is responsible to deliver, which is hard to coordinate when doing regression and acceptance tests and releasing, and thus delay happens.

Also, new feature requests occur during the development cycle, and due to the long development cycle, new feature requests are added as development backlog after the development cycle started. In this case, development tasks can hardly converge until it is closed to release date.

Starting from v5.0, TiDB adopts the release train model. Release train model is a product development model for requirements gathering, analysis, decision making, release, and issue feedback.

Just like a train delivering goods, decisions need to be made about the priority of the goods, destination, arrival time, which train to load on, which carriage, etc., before the train departs.

The benefits of moving to the release train model are:

1. A quicker feedback cycle and users can benefit from features shipped quicker.
2. Predictability for contributors and users:

>     1. Developers and reviewers can decide in advance what release they are aiming for with specific features.

Author: Contributors include developers and reviewers. They are different classification level.

>     2. If a feature misses a release train we have a good idea of when it will show up.

Author: The original expression is better IMO. "release train of the current release" is tweak.

>     2. If a feature misses a release train we have a good idea of when it will show up.

Author: Also "it" means the feature, that the feature will be postponed to be considered in the next release if planned.

>     2. If a feature misses a release train we have a good idea of when it will show up.

Reviewer: For people that are quite familiar with the background, there is no problem using the original expression. But for others, they might have the same question as me: Does "it" in this sentence mean "a feature" or "a release train"?

>     2. If a feature misses a release train we have a good idea of when it will show up.

Author: ```suggestion
    2. If a feature misses a release train, we have a good idea of when the feature will show up later.
```

what about this?

>     2. If a feature misses a release train we have a good idea of when it will show up.

Reviewer: LGTM

    3. Users know when to expect their features

> 3. Transparency. There will be a published cut-off date (AKA code freeze) for the release and people will know about it in advance. Hopefully this will remove the contention around which features make it.

Author: Meaning changes. The original expression focuses that we have code freeze so that features miss the date won't be included. "the contention about which features can meet the date" expresses weirdly.

> 3. Transparency. There will be a published cut-off date (AKA code freeze) for the release and people will know about it in advance. Hopefully this will remove the contention around which features make it.

Reviewer: Similar comment here, one common tip to make the content as clear as possible is using pronouns carefully. For the last sentence, perhaps, we could make it clear that what "this" and "it" represents to avoid confusion and misunderstanding.

> 3. Transparency. There will be a published cut-off date (AKA code freeze) for the release and people will know about it in advance. Hopefully this will remove the contention around which features make it.

Author: see https://github.com/pingcap/tidb-dev-guide/pull/60/files#r676745372

> 4. Quality. we've seen issues pop up in release candidates due to last-minute features that didn't have proper time to bake in. More time between code freeze and release will let us test more, document more and resolve more issues.

Author: Unnecessary changes.

> 5. Project visibility and activity. Having frequent releases improves our visibility and gives the community more opportunities to talk about TiDB.

Author: unnecessary changes

Since nothing is ever perfect, there will be some downsides:

1. Most notably, features that miss the code-freeze date for a release will have to wait few months for the next release train. Features will reach users faster overall as per benefit #1, but individual features that just miss the cut will lose out.
2. For users, figuring out which release to use and having frequent new releases to upgrade to may be a bit confusing.
3. Frequent releases mean we need to do bugfix releases for older branches.

We decided to experiment with release train model and see if the benefits for us as a community exceed the drawbacks. 

> ## How will TiDB development process look like?

Author: unnecessary changes

> At this stage we are planning to make a release every two months.

Author: unnecessary changes

Thus, a typical development cycle takes two months, which we call a sprint. For example, the development cycle of v5.2 is from the end of June to the end of August and called Sprint 4.

> Two weeks before the release date, the release manager will cut branches, publish a list of features that will be included in the release, and also announce code-freeze, after which only fixes for blocking bugs will be merged. This announcement will be posted on the [TiDB Internals forum](https://internals.tidb.io/).

Reviewer: ```suggestion
Two weeks before the release date, the release manager will create a branch for the new release based on the master branch, publish a list of features to be included in the release, and also announce the code-freeze, after which only fixes for blocking bugs can be merged. This announcement will be posted on the [TiDB Internals forum](https://internals.tidb.io/).
```

For release train model, we will strictly ensure that a release happens on a given date. For example, in the release of v5.2, we will decide to have a release by the end of August and we will stick to it. We will drop features that we think will not make it into the release at the time of code freeze and also avoid taking on new features into the release branch. Master development can continue as usual and those features will be in the following release. 

> Ideally, we would have started stabilization around this time. After code freeze, pull requests to the release branch will be merged only if any blocker bugs are identified. In a rare scenario, we could end up with a feature that passed the code freeze bar but still fails to complete on time. Such features will also be dropped from the release at the end to meet the release deadline.

Reviewer: ```suggestion
Ideally, we would have started stabilization once we create the new release branch. After the code-freeze date, only pull requests of blocker bugs can be merged to the new release branch. In a rare scenario, it is possible that few features pass the code freeze bar but still fail to be completed on time. Such features will also be dropped from the release train in the end to meet the release deadline.
```

> ## What happens if features don't complete?

Reviewer: ```suggestion
## What happens to features that are not completed yet?
```

Author: It is a bit confused by "yet". I tend to keep the original expression.

Reviewer: it sounds inappropriate to say " features don't complete", because "feature" is a noun that cannot do things like humans.  How about changing this title to  " What happens if features are not completed"?

Features tend to be of different complexity. Some of them can be implemented within a single release while some span multiple releases. With release train model, we would need to ensure that ongoing features do not affect the stability of the release. There are couple of options -

> 1. Ensure that every feature is broken down into testable units and only testable units get checked in. This means that good set of unit test and system tests are written for sub tasks before they are checked in. This will ensure that master is in a relatively stable state to release at any point of time.

Reviewer: 这句没看懂，这里的 sub-task 是给谁的 sub-task。另外， check in 又是啥意思，是 merge 吗？

> 1. Ensure that every feature is broken down into testable units and only testable units get checked in. This means that good set of unit test and system tests are written for sub tasks before they are checked in. This will ensure that master is in a relatively stable state to release at any point of time.

Author: Check-in is almost merge, I'll update it.

> 1. Ensure that every feature is broken down into testable units and only testable units get checked in. This means that good set of unit test and system tests are written for sub tasks before they are checked in. This will ensure that master is in a relatively stable state to release at any point of time.

Author: Subtasks is of the feature, no matter who it is assigned to. I will keep the content.

> 2. Use feature branches. Create branches from master that are focused on a specific feature. The feature developers ensure that the branch is in sync with master from time to time. Once there is high level of confidence on the stability, it can be merged into master. This approach has the additional overhead of branching and performing merges from time to time.

Reviewer: ```suggestion
2. Use feature branches. For a specific feature, the feature developers create a branch from the master branch and ensure that the branch is in sync with the master branch from time to time. Only when the feature developers and reviewers have a high level of confidence in the feature stability, the feature can be merged into master. This approach brings the additional overhead of branching and performing merges from time to time.
```

> In practice, the right approach would be a mix of both. The feature developers need to make this call depending on the complexity of the feature.

Reviewer: ```suggestion
In practice, the right approach can be a mix of both. The feature developers need to make the decesion depending on the complexity of the feature.
```

Starting from v5.0, TiDB adopts the release train model, which is a product development model for requirements gathering, analysis, decision making, release, and issue feedback.

Just like a train delivering goods, decisions need to be made about the priorities of the goods, destination, arrival time, which train to load on, which carriage, etc., before the train departs.

The benefits of moving to the release train model are as follows:

1. A shorter feedback cycle: users can benefit from features shipped faster.
2. Easier predictability for contributors and users:
    1. Developers and reviewers can decide in advance the target release to deliver specific features.
    2. If a feature misses a release train we have a good idea of when it will show up.
    3. Users know when to expect their features.

> 3. Transparency. There will be a published cut-off date (AKA code freeze) for the release and people will know about it in advance. Hopefully this will remove the contention around which features make it.

Author: ```suggestion
3. Transparency. There will be a published cut-off date (AKA code freeze) for the release and people will know about the date in advance. Hopefully this will remove the contention around which features will be included in the release.
```

what about this @qiancai ?

Reviewer: LGTM
